cxx-tests/src/tests.rs

```diff
@@ -86,30 +109,214 @@ mod bindgenish {
       }
     }
   }
+
+  unsafe impl MoveNew for Foo {
+    unsafe fn move_new(
+      mut src: Pin<moveit::MoveRef<Self>>,
+      this: Pin<&mut MaybeUninit<Self>>,
+    ) {
+      let src: &mut _ = ::std::pin::Pin::into_inner_unchecked(src.as_mut());
+      let this = std::mem::transmute(this);
+      super::ffi::foo_move(this, src);
+      super::ffi::foo_destructor(src);
+    }
+  }
+
+  impl Drop for Foo {
+    fn drop(&mut self) {
+      unsafe { super::ffi::foo_destructor(self) };
+    }
+  }
+
+  #[repr(C)]
+  pub struct Bar {
+    // opaque
+    _pin: PhantomData<PhantomPinned>,
+    _data: [u32; 4],
+  }
+
+  unsafe impl ExternType for Bar {
+    type Id = type_id!("Bar");
+    type Kind = Opaque;
+  }
+
+  unsafe impl MakeCppStorage for Bar {
+    unsafe fn allocate_uninitialized_cpp_storage() -> *mut Self {
+      let bar = super::ffi::bar_create_uninitialized();
+      bar
+    }
+
+    unsafe fn free_uninitialized_cpp_storage(ptr: *mut Self) {
+      super::ffi::bar_free_uninitialized(ptr);
+    }
+  }
+
+  impl Bar {
+    pub fn new() -> impl New<Output = Self> {
+      unsafe {
+        moveit::new::by_raw(|space| {
+          // TODO can we get rid of the transmute?
+          let space = std::mem::transmute(space);
+          super::ffi::bar_constructor(space)
+        })
+      }
+    }
+  }
+
+  unsafe impl MoveNew for Bar {
+    unsafe fn move_new(
+      mut src: Pin<moveit::MoveRef<Self>>,
+      this: Pin<&mut MaybeUninit<Self>>,
+    ) {
+      let src: &mut _ = ::std::pin::Pin::into_inner_unchecked(src.as_mut());
+      let this = std::mem::transmute(this);
+      super::ffi::bar_move(this, src);
+      super::ffi::bar_destructor(src);
+    }
+  }
+
+  impl Drop for Bar {
+    fn drop(&mut self) {
+      unsafe { super::ffi::bar_destructor(self) };
+    }
+  }
+}
+
+/// Queries C++ to find the current state of object lifetimes.
+/// Its main purpose is actually to sneakily enforce a mutex such that
+/// tests do not run in parallel and corrupt that state.
+struct StatusChecker<'a>(MutexGuard<'a, ()>);
+
+impl<'a> StatusChecker<'a> {
+  fn new() -> Self {
+    static MUTEX: OnceCell<Mutex<()>> = OnceCell::new();
+    let guard = MUTEX.get_or_init(|| Mutex::new(())).lock().unwrap();
+    ffi::reset_status();
+    Self(guard)
+  }
+
+  fn assert_status(&self, expected: ffi::Status) {
+    assert_eq!(ffi::get_status(), expected);
+  }
 }
 
 #[test]
 fn test_stack_emplacement() {
-  moveit! {
-    let mut foo = bindgenish::Foo::new();
+  let status_checker = StatusChecker::new();
+  {
+    status_checker.assert_status(ffi::Status::Unallocated);
+    moveit! {
```

Comment:
I encourage the form `moveit!(let ...);` for single-line stuff so I'd like the library to be consistent.